GetAppSession Fix

jorgecuesta opened the pull request:
getAppSession is right now calling tons of times per servicer on new blocks, overloading the fullNode.

jorgecuesta committed:
Added /v1/health endpoint and update rpc spec.

jorgecuesta committed:
Fix #1457; Fix #1484;
Fix a high memory consumption that also is part of the issue #1457.
Under high load of requests (1000/rps or more) the RAM got crazy and scale up to 40GB or close to that.
Now after the fix of #1457 with the worker pool, the node remains under 14gb of ram in my local tests.

jorgecuesta committed:
Initial Geo-Mesh work

jorgecuesta committed:
remove idle timeout in favor to use default (5 seg) because otherwise keep cpu in high load.

jorgecuesta committed:
* Update mesh start time to start at same nodes will.

jorgecuesta committed:
* Added background cache compaction
* Fixed RPC timeout handled as Seconds instead of Milliseconds
* Updated mesh.md to handle new cache configurations
* Updated mesh.md to list /v1/private/mesh/session as required on the whitelist endpoints/paths

jorgecuesta committed:
* Fixed chains hot reload with editor like vim
* Fixed /v1/private/mesh/updatechains to properly update them on memory and disk
* Added hot reload for servicer private key files (add & remove)
  * on add turn on the checks and start allowing it
  * on remove stop receiving and consume all the pending relays in queue
* Version bump

jorgecuesta committed:
* Removed unnecessary initCron call
* Enhanced log about missing sessions
* Version Bump

jorgecuesta committed:
* Fixed issue that allow session to remains on cache forever if the private key is removed after it been supported by the mesh node.
* Version Bump

jorgecuesta committed:
Removed hardcoded values in favor of use GlobalPocketConfig

jorgecuesta committed:
* Removed disk session storage due to issues with size growing (temporal solution)
* Fixed error that panic process when load servicer_url without http/https schema. Now it will properly report the error.
* Added manual cron to compact relays database every hour.
* Removed a log2.Fatal that was crashing the process.

jorgecuesta committed:
Fixed:
* relay_cache_background_sync_interval was not used
* relay_cache_background_compaction_interval was not used

Added:
* hot_reload_interval allow to turn off using 0 the hot reload of chains/servicers - otherwise the amount of MS it will check the files again

Updated:
* Now health check of servicers is done every 60s - was 30s - future: will be configurable through config.json
* Now old sessions are evaluated to be removed every 30m - was 30s - future: will be configurable through config.json
* config.json example of docs.

Removed:
* Manual relays db compaction job removed; We receive reports that it was corrupting relays database if you run at same time of background configured by relay_cache_background_compaction_interval

jorgecuesta committed:
Refactor handle relay to ensure relays are stored in time and deleted from storage in any case after they are success/failed.
Fixed log that was printing node instead of app public key.

jorgecuesta committed:
Fixed issue preventing nodes to track remaining relays properly for every servicer on same session.

jorgecuesta committed:
Update cleanup old session cron time.

jorgecuesta committed:
Remove random selection of servicer because produce error for those running mixed servicers.

jorgecuesta committed:
Update go.mod. Update Servicer version to latest official release RC-0.9.2. Bump Mesh client to RC-0.2.6

jorgecuesta committed:
Update mesh.md

jorgecuesta committed:
Initial rework to speed up bootstrap times of mesh client.
Added different key format.
Refactor connectivity checks.
Refactor node/servicer internal structure of mesh to reduce amount of worker/cron instances.
Refactor chains/keys reload.

jorgecuesta committed:
Fixed minor issues.
Added FullNode worker dynamic resize on servicers change.
Updated servicers reload to only run the modification on maps when there is something new/removed.

jorgecuesta committed:
Replace sync.Map with xsync.Map to use Generic in favor of performance and better readability of the code without so many casts.
Refactor fullNode.Servicer to be a map instead of a slice.
Enhance a bit more the logs and bootstrap time information.

jorgecuesta committed:
Added worker metrics.
Added metrics config support.
Refactor code to split in files.
Bump pond version to 1.8.3 (patch).
Clean up the code.

jorgecuesta committed:
Update docs.
Update config to handle rpc timeout for different things like chains, client and pocket node calls with a different value.

jorgecuesta committed:
Fixed metrics and cyclic dependencies.

jorgecuesta committed:
Fixed a bug that disallow the ability to reuse http 1.x connections.

jorgecuesta committed:
Added support to Chains & Servicer http client options to be configurable by config file.

jorgecuesta committed:
Updated rpc-spec.yaml.
Ensure that http response body is read even on errored request to reuse connections.

jorgecuesta committed:
Fixed issue with /v1/private/mesh/updatechains endpoint.
Enhanced chains reload logs.
Enhanced startup logs.

jorgecuesta committed:
Added specialized metrics for mesh instead of reuse the pocket client.

jorgecuesta committed:
Removed dynamic capacity on servicer worker due to the need of handle so many edge cases and possible infinite goroutine spams.
Added name property to nodes as optional key, if not set use the hostname of the node url.
Added minWorker, maxWorker, maxCapacity to prometheus metrics collectors.
Refactor minWorker, maxWorker and maxCapacity option in config.
Bump default to a more real world value.
Updated docs.

jorgecuesta committed:
Updated docs.

jorgecuesta committed:
Added option to allow debug logs omit request/response payloads.

jorgecuesta committed:
Added json schema support to additional properties. Those will not impact on mesh code, but help node runners to keep internal track.

jorgecuesta committed:
Rework metrics to handle less metrics with more labels. Enhance group/queries on prometheus.
Added chains name map so those metrics could contain the chain name you wish.
ChainsNameMap could work with a local file or remote endpoint (GET)

jorgecuesta committed:
Set chains_name_map default as empty so avoid to read it and write an error log. Does not affect the code but is unnecessary.

jorgecuesta committed:
Added metrics_moniker.
Added status_type and status_code labels to error metrics.
Added internal, notify and chain error metrics.

jorgecuesta committed:
Fixed issue with error metrics

jorgecuesta committed:
Rename status_label to status_code.
Added metrics docs and basic geo-mesh grafana dashboard.

jorgecuesta committed:
Moved authtoken to Authorization header to avoid the security token could be leaked on logs.
Moved Grafana Dashboard to a file to easily compare/copy from raw github files.
Updated mesh.md

jorgecuesta committed:
Enhance code based on blade suggestions.
Enhance docs based on blade suggestions.

jorgecuesta committed:
Added golang-set dependency link to docs.

jorgecuesta committed:
Added missing chain request log.

jorgecuesta committed:
Added support to configure more http client that handle mesh request and the one dispatching to chains.

jorgecuesta committed:
Fixed issue with log chain requests.

jorgecuesta committed:
Allow to remove non-printable characters from chain path

jorgecuesta committed:
Updated default values.

jorgecuesta committed:
Updated mesh.md

jorgecuesta committed:
Updated mesh.md and default value of chain_request_path_cleanup

jorgecuesta committed:
Merge pull request #1 from pokt-scan/initial-enhancement

## High-Level Summary
This update includes changes that allow for faster boot time and better overall performance & maintainability.

* Mesh bootstrap time has been greatly reduced to almost 0, which improves overall performance and makes it easier to maintain your key files.
* Mesh now supports hot reloading of keys and chains efficiently by deduplicating full nodes, which allows for faster updates and better overall performance.
* Code has been split into smaller files to make it easier to navigate and modify, which helps with both development and maintenance.
* The Relay, Relay Time, and Error metrics now use labels for chain id and name to differentiate between different chains and to help with identifying and categorizing the various metrics.
* Misc bug fixes such as connection re-use and invalid sessions which should increase the performance of your mesh and full node.

### Breaking Changes
This changes require you update both: Pocket Node & Mesh Node

1. Added `/v1/private/mesh/check` endpoint
2. Moved `/v1/mesh/health` to `/v1/private/mesh/health`
3. Removed `/v1/private/mesh/servicer`
4. Removed Prometheus metrics using chain id as part of the metric name
5. Removed `authtoken` query param from private mesh endpoints in favor of `Authorization: <token>` header.
    *  **_NOTE: No changes need on your side unless you have some tooling built over `/v1/private/mesh/*` endpoints_**

**_NOTE:_** 

- You must add the new private endpoints (`check`) and use the same code/image on both sides or your mesh node will not start.
- You could remove the `/v1/health` endpoint from your proxy.


### Configuration Changes
1. Added `node_check_interval` to configure the rate at a full Node is health checked.
2. Added `log_chain_request` and `log_chain_response` to avoid unnecessary debug logs.
3. Added `chains_name_map` and `remote_chains_name_map` to enhance exposed metrics with chain name.
4. Added `metrics_moniker` to help to identify a mesh node instance on metrics queries.
7. Added `metrics_report_interval` to configure the rate at metrics about workers are reported.
8. Added `<servicer|chain>_rpc_max_idle_connections`
9. Added `<servicer|chain>_rpc_max_conns_per_host`
10. Added `<servicer|chain>_rpc_max_idle_conns_per_host`
11. Added `chain_drop_connections`
12. Added `client_rpc_read_timeout`
13. Added `client_rpc_read_header_timeout`
14. Added `client_rpc_write_timeout`
15. Added `chain_request_path_cleanup`
16. Removed `hot_reload_interval` in favor of `keys_hot_reload_interval` and `chains_hot_reload_interval` as separated values
17. Rename worker pool options to specify a different set for Servicer or Metrics
18. Remove `rpc_timeout` in favor of: `client_rpc_timeout` and `chain_rpc_timeout`

### Core Changes
#### New:
1. New keys structure
2. Deduplication of nodes (reduce bandwidth usage and faster boot time)
3. Allow control RPS of mesh to a node (using servicer_max_workers) due to how workers are used now
4. Chains check. Check against the Pocket node if the chains id on the chain file matches the one on Pocket Node.
6. Added a Worker for metrics
7. Chains name maps (local or remote) to enhance metrics
8. Added endpoint to allow mesh node check against servicer node (health, chains, addresses)
9. Added sanitization of the URL before call chain due to errors observed on chains like Avax/DFK sending `\t` characters on the path.

#### Fix:
1. Invalid sessions error
2. HTTP connection drops due to avoiding read response body (decreases CPU utilization and bandwidth)
3. Handle a few cases where after an error it keeps going without interrupting the flow, invalidating a relay/session.
4. Update chains using /v1/private/mesh/updatechains endpoint

#### Rework:
1. Relay, Relay Time, and Error metrics now use labels for chain id and name
2. Split Keys and Chains hot reload interval to allow them to work independently.
3. Moved the `authtoken` query parameter used to call private method to be passed using the HTTP header `Authorization` to avoid it being leaked on logs.
4. Split code in files to help development and readability of it.
5. Bootstrap time reduce to almost 0

#### Dependencies:
1. Added [xsync](github.com/puzpuzpuz/xsync) to have better performance that the native for concurrent access.
2. Added [gojsonschema](github.com/xeipuuv/gojsonschema) to allow validate keys and chains name map format.
3. Added [golang-set](github.com/deckarep/golang-set/v2) to manage set of values in arrays faster and async safe
4. Bump version github.com/alitto/pond from v1.8.1 to v1.8.3
5. Bump version github.com/prometheus/client_golang from v1.11.0 to v1.11.1

### Bump Version
New version: RC-0.3.0

### Docs
1. Update mesh.doc
2. Added missing links to dockerhub
3. Added Grafana dashboard to consume the new metrics.
4. Updated rpc-spec.yaml

jorgecuesta committed:
Renamed invalidate session codes.
Added non retryable relays codes.
Change retryable handler to use a dedicated function for it.

jorgecuesta committed:
Bump version

jorgecuesta committed:
Fixed issue on report metrics without proper return http code from blockchain call.
Reduce default values for the retries.
Removed CodeRequestHash & CodeInvalidBlockHeightError from invalidate session codes.

jorgecuesta committed:
Fixed issue on report metrics without proper return http code from blockchain call.
Reduce default values for the retries.
Moved CodeRequestHash & CodeInvalidBlockHeightError from invalidate session codes to non retryable ones.

jorgecuesta committed:
Added support to lean node for evidence worker. Now is one worker per address. Prevent the race condition on the evidence store and do not hold relays evidence due to other address is using the queue.

jorgecuesta committed:
Remove chain url from the error message returned as part of the response. Instead, return a more generic information but log the real error for internal node runner tracking.
Enhance error status types for servicer and chain not found.

jorgecuesta committed:
Initial work over session fix. WIP on unit tests (learning how to)

nodiesBlade commented on app/cmd/rpc/mesh/fullnode.go:
```diff
@@ -249,7 +267,10 @@ func createNode(urlStr, name string) *fullNode {
-		Crons:     nodeCronJobsWorker,
+		// just a default because on each check request the node will return the right value for the height returned
+		// TODO: implement a default from config as "fallback" before get the real node value
+		BlocksPerSession: 4,
```
I think you can retrieve it from /allparams endpoint?

nodiesBlade replied:
/v1/query/allparams

jorgecuesta replied:
that is added to the check endpoint so we do not need another call to node, reuse the same, that one is a "safety" default value in case it is required before we get the first status check from node. It should not be required because the connectivity check run sync before mesh start servicing.

nodiesBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -44,19 +47,14 @@ type DispatchResponse struct {
 		return false
 	}
 	// loop over the nodes
 	for _, node := range sn.Session.Nodes {
```
converting this to a map would be more efficient in terms of CPU (with expense of slightly more memory) if you store it in the dispatch response.

though it is a premature optimization given it's only 24 nodes

jorgecuesta replied:
yeah and also that is reusing the same structure fullNode return from pocket core client basics.

nodiesBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -89,124 +87,494 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+	)
+	jsonData, e1 := json.Marshal(payload)
+	if e1 != nil {
+		statusCode = -2
```
custom status codes are hard to maintain if you just do -1 to -5, use `iota`

https://www.gopherguides.com/articles/how-to-use-iota-in-golang
```
const (
	Red int = iota
	Orange
	Yellow
	Green
	Blue
	Indigo
	Violet
)
```

jorgecuesta replied:
100% thanks for the link, will implement asap.

nodiesBlade commented on app/cmd/rpc/mesh.go:
```diff
@@ -383,7 +383,29 @@ func meshServicerNodeCheck(w http.ResponseWriter, r *http.Request, ps httprouter
+		return
+	}
+
+	paramReturn, err := app.PCA.QueryParam(latestHeight, "pos/BlocksPerSession")
```
Ah, i see where you made the change now

nodiesBlade replied:
shouldn't you be getting the blocks per session for the "dispatch relay height", not the latest height?

jorgecuesta replied:
this is the "check" endpoint, so always return the "latest" height + block per session, I understand what u say but we are getting "latest" information on this endpoint.

jorgecuesta replied:
> dispatch relay height

for this we could been on a gap when we are talking about future sessions and we will not be asking for older sessions, because they will not be handled by the node, so always talking about "current" session (like the last one) + future sessions, that we cant be sure they are "real" so we just trust on what it is, and use last "known" blockPerSession for those "future" sessions

nodiesBlade replied:
I think we will need to revisit this (and maybe even Pocket core) if the network ever decides to adjust BlocksPerSession. It isn't something the network should do carelessly, and third party clients will need to adjust the behavior as well (i.e should blocks per session value be reflected before, during, or after the end of session). Anyhow there isn't immediate need to anything here, as long as we both keep it in mind

jorgecuesta replied:
Yeah you are right, hopefully this will not be touched until v1 or not even there. I think for now is safe use a default and also ask the node for the value on the latest block, because that work for the "future" session and for the current one.

nodiesBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -89,124 +87,494 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
-// getAppSession - call ServicerURL to get an application session using retrieve header
-func getAppSession(relay *pocketTypes.Relay, model interface{}) *SdkErrorResponse {
-	servicerNode := getServicerFromPubKey(relay.Proof.ServicerPubKey)
+func (ns *NodeSession) CountRelay() (*NodeSession, bool) {
```
Why would you ever need to return the node session itself?

jorgecuesta replied:
concat functions? like sugar session just that.

jorgecuesta replied:
good one, cleaning it.

nodiesBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -89,124 +87,494 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+}
+
+func (ns *NodeSession) ReScheduleValidationTask(session *Session, servicerPubKey string) {
+	if ns.RetryTimes > 20 {
```
Set 20 as a const to avoid magic numbers
https://en.wikipedia.org/wiki/Magic_number_(programming)

jorgecuesta replied:
thanks will fix, good catch

jorgecuesta replied:
I set this to a default value. `SessionStorageValidateRetryMaxTimes`

jorgecuesta committed:
wip on add test to session storage.

jorgecuesta committed:
Fixed issue that was forcing retry relays notifications when they should be discarded.
Added session validation before notify to avoid call a node if the session was already invalidated.
Bump version to 0.3.2

jorgecuesta committed:
Enhance code base on @PoktBlade comments.

jorgecuesta committed:
Rebase on latest geo-mesh commit.
Enhance code base on @PoktBlade comments.

nodiesBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -89,124 +94,497 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+	)
+
+	ns.IsValid = false
+	ns.Error = NewSdkErrorFromPocketSdkError(pocketTypes.NewOverServiceError(ModuleName))
```
should return maximum relays sealed not overservice (code 90)

jorgecuesta replied:
why? when the evidence can take more relays, should be OverService

nodiesBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -89,124 +94,497 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+	fullNodeHeight := servicerNode.Status.Height
+	blocksPerSession := servicerNode.BlocksPerSession
+	return (sessionBlockHeight >= fullNodeHeight &&
+		(fullNodeHeight%blocksPerSession == 0 || fullNodeHeight%blocksPerSession == 1)) &&
```
create new var

fullNodeSessionBlockHeight = fullNodeHeight % blocksPerSession and then your comparison can be as simple as 

fullNodeSessionBlockHeight  <= 1

PoktBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -89,124 +94,497 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+	blocksPerSession := servicerNode.BlocksPerSession
+	return (sessionBlockHeight >= fullNodeHeight &&
+		(fullNodeHeight%blocksPerSession == 0 || fullNodeHeight%blocksPerSession == 1)) &&
+		relay.Proof.SessionBlockHeight-servicerNode.GetLatestSessionBlockHeight() <= 1
```
Actually, do you really even need the above statement if you have this one?

nodiesBlade commented on app/cmd/rpc/mesh/session.go:
```diff
@@ -89,124 +94,497 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+	blocksPerSession := servicerNode.BlocksPerSession
+	return (sessionBlockHeight >= fullNodeHeight &&
+		(fullNodeHeight%blocksPerSession == 0 || fullNodeHeight%blocksPerSession == 1)) &&
+		relay.Proof.SessionBlockHeight-servicerNode.GetLatestSessionBlockHeight() <= 1
```
Assume session roll over:

1. Servicer on Height 9 (sessionBlockHeight 9)
2. Dispatcher on Height 8 (sessionBlockHeight 5)

the difference here should be `<= blocksPerSession` not 1. We are bounding it by start of session, correct?

nodiesBlade replied:
comments on the behavior of this function would be nice

nodiesBlade replied:
```fullNodeHeight%blocksPerSession == 0 || fullNodeHeight%blocksPerSession == 1```

wondering if this should be == 3 (aka at the last block of the session, i.e block 8), not <= 1

jorgecuesta replied:
@PoktBlade check it now, also uploaded the unit test for it.

jorgecuesta committed:
Rework ShouldAssumeOptimisticSession function.
Added unit-test for ShouldAssumeOptimisticSession function.

jorgecuesta committed:
Fixed and issue that provoke session not behind found.

jorgecuesta committed:
Added missing blockchain validation on relays.
Added missing session height validation on relays.

jorgecuesta committed:
Bump Version to BETA-0.4.0

jorgecuesta committed:
Added "current" session to the same logic of the "optimistic" one to reduce the latency on session changes.
Fixed issue on /v1/private/mesh/check that throw error asking latest height params.
Fixed relay_time metric that was not wrapping around "validate" method, and it is the one that could take more time after call blockchain.
Added missing initialization of the metrics for session storage.
Added metric for call blockchain as a separated one.
Enhance few error logs.
Added additional metrics.
Fixed multiple issues detected around offloaded session validate tasks.
Fixed an issue that prevent session behind validated after the session already exists for other servicer handled by the mesh.
Added more defensive code question to avoid panic due to nil pointers.

jorgecuesta committed:
Refactored session.go to handle session in a single map to avoid race conditions and reduce memory consumption.
Enhance logs for relay and session.
Enhance metrics for notify process.

jorgecuesta committed:
remove unnecessary log. fix session_test.go
